_posts/2019-09-30-bert-r/bert_r.Rmd

```diff
@@ -0,0 +1,536 @@
+
+## Model structure
+
+Loading a pre-trained BERT model is straightforward. [Downloaded zip file contains](https://storage.googleapis.com/bert_models/2018_10_18/uncased_L-12_H-768_A-12.zip):
+
+- *bert_model.ckpt* which is for loading the weights from the TensorFlow checkpoint
+- *bert_config.json* which is a configuration file
+- *vocab.txt* which is for text tokenization
+
+```{r}
+pretrained_path = '/Users/turgutabdullayev/Downloads/uncased_L-12_H-768_A-12'
+config_path = file.path(pretrained_path, 'bert_config.json')
+checkpoint_path = file.path(pretrained_path, 'bert_model.ckpt')
+vocab_path = file.path(pretrained_path, 'vocab.txt')
+```
+
+## Import Keras-Bert module via reticulate
+
+Let's upload keras-bert via reticulate and prepare tokenizer object. BERT tokenizer will help us to turn words into indices.
+
+```{r}
+k_bert = import('keras_bert')
+token_dict = k_bert$load_vocabulary(vocab_path)
+tokenizer = k_bert$Tokenizer(token_dict)
+```
+
+### How tokenizer works?
+
+BERT uses a WordPiece tokenization strategy. If a word is Out-of-vocabulary (OOV), then BERT will break it down into subwords. (eating => eat, ##ing).
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset", fig.cap = "[BERT input representation. The input embeddings are the sum of the token embeddings, the segmentation embeddings, and the position embeddings](https://arxiv.org/pdf/1810.04805.pdf)"}
+knitr::include_graphics("images/emb.png")
+```
+
+## Define model parameters and column names
+
+Similar to Keras standard parameters: input length, batch size, epochs and learning rate should be defined for training a BERT model with Keras.
+
+```{r}
+seq_length = 50L
+bch_size = 70
+epochs = 1
+learning_rate = 1e-4
+
+DATA_COLUMN = 'comment_text'
+LABEL_COLUMN = 'target'
+```
+
+> Note: the max input length is 512 and it is extremely computational even for GPU. 
+
+## Load BERT model into R
+
+We can load the BERT model and automatically pad sequences with ```seq_len``` function. This is why using keras-bert^[[Implementation of the BERT. Official pre-trained models could be loaded for feature extraction and prediction](https://github.com/CyberZHG/keras-bert)] is a better choice, it makes the loading process very easy and comfortable.
```

Comment:
The comparison here comes a bit out of the blue (better than what?)
I know what you mean, but most readers won't. I'd just leave that out (you can still emphasize how comfortable the process is.)

Reply:
I just removed comparison.